br/pkg/utils/worker.go

```diff
@@ -61,41 +63,54 @@ func (pool *WorkerPool) Apply(fn taskFunc) {
 
 // ApplyWithID execute a task and provides it with the worker ID.
 func (pool *WorkerPool) ApplyWithID(fn identifiedTaskFunc) {
-	worker := pool.ApplyWorker()
+	worker, _ := pool.ApplyWorker(context.TODO())
 	go func() {
 		defer pool.RecycleWorker(worker)
 		fn(worker.ID)
 	}()
 }
 
 // ApplyOnErrorGroup executes a task in an errorgroup.
-func (pool *WorkerPool) ApplyOnErrorGroup(eg *errgroup.Group, fn func() error) {
-	worker := pool.ApplyWorker()
+func (pool *WorkerPool) ApplyOnErrorGroup(ctx context.Context, eg *errgroup.Group, fn func() error) error {
+	worker, err := pool.ApplyWorker(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
 	eg.Go(func() error {
 		defer pool.RecycleWorker(worker)
 		return fn()
 	})
+	return nil
 }
 
 // ApplyWithIDInErrorGroup executes a task in an errorgroup and provides it with the worker ID.
-func (pool *WorkerPool) ApplyWithIDInErrorGroup(eg *errgroup.Group, fn func(id uint64) error) {
-	worker := pool.ApplyWorker()
+func (pool *WorkerPool) ApplyWithIDInErrorGroup(ctx context.Context, eg *errgroup.Group, fn func(id uint64) error) error {
+	worker, err := pool.ApplyWorker(ctx)
+	if err != nil {
+		return errors.Trace(err)
+	}
 	eg.Go(func() error {
 		defer pool.RecycleWorker(worker)
 		return fn(worker.ID)
 	})
+	return nil
 }
 
 // ApplyWorker apply a worker.
-func (pool *WorkerPool) ApplyWorker() *Worker {
+func (pool *WorkerPool) ApplyWorker(ctx context.Context) (*Worker, error) {
 	var worker *Worker
+	if ctx.Err() != nil {
+		return nil, context.Cause(ctx)
```

Comment:
we still return "context cancel" error and how does it solve origin issue?

Reply:
If the function in the workerpool return an error, then the errgroup `eg` would cancel the context `ectx` with the error.
Therefore the context `ectx` has these values:
```
ectx: context.Context {
  Error: "context cancel",
  Cause: "{error}",
}
```